illumos-utils/src/running_zone.rs

```diff
@@ -169,13 +174,39 @@ mod zenter {
         fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
         fn ct_tmpl_activate(fd: c_int) -> c_int;
         fn ct_tmpl_clear(fd: c_int) -> c_int;
+        fn ct_ctl_abandon(fd: c_int) -> c_int;
     }
 
     #[link(name = "c")]
     extern "C" {
         pub fn zone_enter(zid: zoneid_t) -> c_int;
     }
 
+    pub fn get_contract(pid: pid_t) -> std::io::Result<ctid_t> {
+        // The offset of "id_t pr_contract" in struct psinfo which is an
+        // interface documented in proc(5).
+        const CONTRACT_OFFSET: u64 = 280;
+
+        let path = format!("/proc/{}/psinfo", pid);
+
+        let mut file = File::open(path)?;
+        file.seek(SeekFrom::Start(CONTRACT_OFFSET))?;
+        let mut buffer = [0; 4];
+        file.read_exact(&mut buffer)?;
+        Ok(ctid_t::from_ne_bytes(buffer))
+    }
+
+    pub fn abandon_contract(ctid: ctid_t) {
+        let path = format!("/proc/{}/contracts/{}/ctl", process::id(), ctid);
+
+        let cpath = CString::new(path).unwrap();
+        let fd = unsafe { libc::open(cpath.as_ptr(), libc::O_WRONLY) };
+        if fd >= 0 {
+            unsafe { ct_ctl_abandon(fd) };
+            unsafe { libc::close(fd) };
+        }
+    }
```

Comment:
This routine should be fallible.  If `open()` fails, we need to return some detail about why.  We should also check the return values of `ct_ctl_abandon()` and `close()` and fail with error reports if those are not as expected.

Any error here represents a serious misunderstanding of the construction of the system, and could lead to more leaked contracts or leaked file descriptors.  At a minimum we need a noisy error log message describing what went wrong; ideally we'd abort the process and get a core file.

Reply:
As both something which should never happen and something which is survivable, I made abandoning the contract best effort. I will adjust so that we get error messages if this occurs but will stop short of panicking.